### PR TITLE
Create the TransactionsProof request

### PR DESCRIPTION
**Pull request by viquezclaudio**
Create a new request, in order to handle Transaction inclusion proofs.

This is necessary for web-clients to request transaction inclusion proofs from history nodes.

### CODE REVIEW

**Comment by sisou** [comment]
Will there also be a method on the consensus to request the proof of a specific transaction? Or does that come later in a separate PR after this one merges the message type only?

**Comment by viquezclaudio** [comment]
That will be part of this PR too, so that the requests can be performed from JS

File: blockchain/src/history/history_store.rs [hunk 1]
```diff
@@ -17,11 +16,12 @@ use nimiq_mmr::mmr::proof::RangeProof;
 use nimiq_mmr::mmr::MerkleMountainRange;
 use nimiq_mmr::store::memory::MemoryStore;
 use nimiq_primitives::policy::Policy;
+use nimiq_transaction::extended_transaction::{ExtTxData, ExtendedTransaction};
+use nimiq_transaction::inherent::InherentType;
```
Comment (jsdanielh) [inline blockchain/src/history/history_store.rs]:
```suggestion
use nimiq_transaction::{extended_transaction::{ExtTxData, ExtendedTransaction}, inherent::InherentType};
```

File: blockchain/src/history/history_store.rs [hunk 2]
```diff
@@ -17,11 +16,12 @@ use nimiq_mmr::mmr::proof::RangeProof;
...
 
 use crate::history::mmr_store::MMRStore;
 use crate::history::ordered_hash::OrderedHash;
-use crate::history::{ExtendedTransaction, HistoryTreeChunk, HistoryTreeProof};
-use crate::ExtTxData;
+use crate::history::{HistoryTreeChunk, HistoryTreeProof};
```
Comment (jsdanielh) [inline blockchain/src/history/history_store.rs]: I'd collapse all of the `crate::history` uses in a single statement

File: consensus/src/messages/handlers.rs [hunk 3]
```diff
@@ -300,3 +300,64 @@ impl Handle<ResponseChunk, Arc<RwLock<Blockchain>>> for RequestChunk {
...
+
+        let proof = blockchain
+            .history_store
+            .prove(self.epoch_number, hashes, None);
+
+        let block_height = self.epoch_number * Policy::blocks_per_epoch();
+
+        let block = match blockchain
+            .chain_store
+            .get_block_at(block_height, false, None)
```
Comment (jsdanielh) [inline consensus/src/messages/handlers.rs:318]: Sounds like we could use:
```
.ok().and_then(|block| { if block.is_macro() {
  // We expect a macro block
    Some(block)
  } else {
    None
}}
```

File: consensus/src/messages/handlers.rs [hunk 4]
```diff
@@ -300,3 +300,64 @@ impl Handle<ResponseChunk, Arc<RwLock<Blockchain>>> for RequestChunk {
...
+            &self.address,
+            self.max.unwrap_or(500),
+            None,
+        );
+
+        let mut transactions = vec![];
+
+        for hash in tx_hashes {
+            // Get all the extended transactions that correspond to this hash.
+            let mut extended_tx_vec = blockchain.history_store.get_ext_tx_by_hash(&hash, None);
```
Comment (jsdanielh) [inline consensus/src/messages/handlers.rs]: Why is this `mut`?

Comment (jsdanielh) [reply consensus/src/messages/handlers.rs]: Couldn't we have done something like:
`transactions.extend(blockchain.history_store.get_ext_tx_by_hash(&hash, None))`
?

File: consensus/src/messages/mod.rs [hunk 5]
```diff
@@ -3,9 +3,11 @@ use std::fmt::{Debug, Formatter};
 use beserial::{Deserialize, Serialize};
 use nimiq_block::{Block, MacroBlock};
 #[cfg(feature = "full")]
-use nimiq_blockchain::HistoryTreeChunk;
+use nimiq_blockchain::{HistoryTreeChunk, HistoryTreeProof};
 use nimiq_hash::Blake2bHash;
+use nimiq_keys::Address;
```
Comment (jsdanielh) [inline consensus/src/messages/mod.rs:9]:
```suggestion
#[cfg(feature = "full")]
use nimiq_keys::Address;
```

File: consensus/src/messages/mod.rs [hunk 6]
```diff
@@ -3,9 +3,11 @@ use std::fmt::{Debug, Formatter};
...
 use nimiq_network_interface::request::{RequestCommon, RequestMarker};
+use nimiq_transaction::extended_transaction::ExtendedTransaction;
```
Comment (jsdanielh) [inline consensus/src/messages/mod.rs:12]:
```suggestion
#[cfg(feature = "full")]
use nimiq_transaction::extended_transaction::ExtendedTransaction;
```

File: consensus/src/messages/mod.rs [hunk 7]
```diff
@@ -202,3 +208,48 @@ impl RequestCommon for RequestHead {
...
+pub struct RequestTransactionsProof {
+    #[beserial(len_type(u16, limit = 128))]
+    hashes: Vec<Blake2bHash>,
+    epoch_number: u32,
+}
+
+#[cfg(feature = "full")]
+impl RequestCommon for RequestTransactionsProof {
+    type Kind = RequestMarker;
+    const TYPE_ID: u16 = 212;
```
Comment (jsdanielh) [inline consensus/src/messages/mod.rs]: 212 is already occupied by `RequestChunk`

File: consensus/src/sync/history/cluster.rs [hunk 8]
```diff
@@ -8,11 +8,12 @@ use std::task::{Context, Poll};
 use futures::{FutureExt, Stream, StreamExt};
 use lazy_static::lazy_static;
 use nimiq_blockchain_interface::{AbstractBlockchain, PushError, PushResult};
+use nimiq_transaction::extended_transaction::ExtendedTransaction;
```
Comment (jsdanielh) [inline consensus/src/sync/history/cluster.rs]: I'd place `nimiq` uses in the next section

File: primitives/account/src/staking_contract/traits.rs [hunk 9]
```diff
@@ -8,6 +8,7 @@ use nimiq_primitives::{account::AccountError, coin::Coin, policy::Policy, slots:
 use nimiq_transaction::account::staking_contract::{
     IncomingStakingTransactionData, OutgoingStakingTransactionProof,
 };
+use nimiq_transaction::inherent::{Inherent, InherentType};
 use nimiq_transaction::Transaction;
```
Comment (jsdanielh) [inline primitives/account/src/staking_contract/traits.rs]:
```suggestion
use nimiq_transaction::{account::staking_contract::{
    IncomingStakingTransactionData, OutgoingStakingTransactionProof,
}, inherent::{Inherent, InherentType}, Transaction};
```

File: primitives/account/tests/accounts.rs [hunk 10]
```diff
@@ -1,13 +1,11 @@
+use rand::{rngs::StdRng, SeedableRng};
 use std::convert::TryFrom;
 use std::time::Instant;
-
-use rand::{rngs::StdRng, SeedableRng};
```
Comment (jsdanielh) [inline primitives/account/tests/accounts.rs:4]: This was actually ok, we list `std` imports apart

File: primitives/transaction/Cargo.toml [hunk 11]
```diff
@@ -20,12 +20,16 @@ strum_macros = "0.24"
 thiserror = "1.0"
 
 beserial = { path = "../../beserial", features = ["derive"] }
+
```
Comment (jsdanielh) [inline primitives/transaction/Cargo.toml]: This line is unnecessary: `beserial` is an internal crate as well

File: primitives/transaction/Cargo.toml [hunk 12]
```diff
@@ -20,12 +20,16 @@ strum_macros = "0.24"
...
 nimiq-bls = { path = "../../bls", features = ["serde-derive"] }
+nimiq-database-value = { path = "../../database/database-value"}
 nimiq-hash = { path = "../../hash", features = ["serde-derive"] }
 nimiq-keys = { path = "../../keys", features = ["serde-derive"] }
 nimiq-macros = { path = "../../macros" }
+nimiq-mmr = { path = "../mmr" }
 nimiq-network-interface = { path = "../../network-interface" }
 nimiq-primitives = { path = "..", features = ["account", "coin", "networks", "policy", "serde-derive"] }
+
```
Comment (jsdanielh) [inline primitives/transaction/Cargo.toml]: This line is not needed as well

File: primitives/transaction/src/extended_transaction.rs [hunk 13]
```diff
@@ -3,15 +3,17 @@ use std::fmt;
 use std::io;
 
 use beserial::{Deserialize, ReadBytesExt, Serialize, SerializingError, WriteBytesExt};
-use nimiq_account::{Inherent, InherentType};
+
```
Comment (jsdanielh) [inline primitives/transaction/src/extended_transaction.rs]: This line is unnecessary: `beserial` is an internal crate

File: primitives/transaction/src/inherent.rs [hunk 14]
```diff
@@ -4,10 +4,13 @@ use beserial::{Deserialize, ReadBytesExt, Serialize, SerializingError, WriteByte
...
 use nimiq_keys::Address;
 use nimiq_primitives::coin::Coin;
-use nimiq_transaction::reward::RewardTransaction;
+
+use crate::reward::RewardTransaction;
 
 #[derive(Clone, Debug, Eq, PartialEq, Copy, Serialize, Deserialize)]
 #[repr(u8)]
+/// Enum that represents the different types of inherents.
+/// They moment when those inherents are applied depends upon the specific type of inherent.
```
Comment (jsdanielh) [inline primitives/transaction/src/inherent.rs]:
```suggestion
/// The moment when these inherents are applied depends upon the specific type of inherent.
```

File: primitives/transaction/src/inherent.rs [hunk 15]
```diff
@@ -28,6 +31,9 @@ impl InherentType {
 }
 
 #[derive(Clone, Debug, Eq, PartialEq)]
+/// An inherent is a special kind of transaction
+/// It does not have a sender account, but it does have a target address and a value that is transfered
+/// For instance, a reward (for producing blocks) is an inherent
```
Comment (jsdanielh) [inline primitives/transaction/src/inherent.rs:36]: I'd add member `rustdoc` as well

File: primitives/transaction/src/inherent.rs (hunk 14 above)
Comment (jsdanielh) [reply primitives/transaction/src/inherent.rs]: I'd add member `rustdoc` as well

**Review by jsdanielh — CHANGES_REQUESTED** [review]

**Review by jsdanielh — APPROVED** [review]

**Pull request merged**